data/transactions/logic/eval.go

```diff
@@ -0,0 +1,1057 @@
+	GroupIndex int
+
+	// pseudo random number generator, or seed parts
+	Source   Source64
+	Seed     []byte
+	MoreSeed []byte
+}
+
+type evalContext struct {
+	EvalParams
+
+	stack   []stackValue
+	program []byte // txn.Lsig.Logic ?
+	pc      int
+	nextpc  int
+	err     error
+	intc    []uint64
+	bytec   [][]byte
+	version uint64
+	scratch [256]stackValue
+
+	checkStack []StackType
+
+	// Ordered set of pc values that a branch could go to.
+	// If Check pc skips a target, the source branch was invalid!
+	branchTargets []int
+}
+
+type opFunc func(cx *evalContext)
+
+// StackType describes the type of a value on the operand stack
+type StackType byte
+
+// StackNone in an OpSpec shows that the op pops or yields nothing
+const StackNone StackType = 0
+
+// StackAny in an OpSpec shows that the op pops or yield any type
+const StackAny StackType = 1
+
+// StackUint64 in an OpSpec shows that the op pops or yields a uint64
+const StackUint64 StackType = 2
+
+// StackBytes in an OpSpec shows that the op pops or yields a []byte
+const StackBytes StackType = 3
+
+func (st StackType) String() string {
+	switch st {
+	case StackNone:
+		return "None"
+	case StackAny:
+		return "any"
+	case StackUint64:
+		return "uint64"
+	case StackBytes:
+		return "[]byte"
+	}
+	return "internal error, unknown type"
+}
+
+// Eval checks to see if a transaction passes logic
+// A program passes succesfully if it finishes with one int element on the stack that is non-zero.
+func Eval(program []byte, params EvalParams) (pass bool, err error) {
+	var cx evalContext
+	version, vlen := binary.Uvarint(program)
+	if version > EvalMaxVersion {
+		cx.err = fmt.Errorf("program version %d greater than max supported version %d", version, EvalMaxVersion)
+		return false, cx.err
+	}
+	if (params.Proto != nil) && (version > params.Proto.LogicSigVersion) {
+		cx.err = fmt.Errorf("program version %d greater than protocol supported version %d", version, params.Proto.LogicSigVersion)
+		return false, cx.err
+	}
+	// TODO: if EvalMaxVersion > version, ensure that inaccessible
+	// fields as of the program's version are zero or other
+	// default value so that no one is hiding unexpected
+	// operations from an old program.
+	cx.version = version
+	cx.pc = vlen
+	cx.EvalParams = params
+	cx.stack = make([]stackValue, 0, 10)
+	cx.program = program
+	for (cx.err == nil) && (cx.pc < len(cx.program)) {
+		cx.step()
+	}
+	if cx.err != nil {
+		if cx.Trace != nil {
+			fmt.Fprintf(cx.Trace, "%3d %s\n", cx.pc, cx.err)
+		}
+		return false, cx.err
+	}
+	if len(cx.stack) != 1 {
+		if cx.Trace != nil {
+			fmt.Fprintf(cx.Trace, "end stack:\n")
+			for i, sv := range cx.stack {
+				fmt.Fprintf(cx.Trace, "[%d] %s\n", i, sv.String())
+			}
+		}
+		return false, fmt.Errorf("stack len is %d instead of 1", len(cx.stack))
+	}
+	if cx.stack[0].Bytes != nil {
+		return false, errors.New("stack finished with bytes not int")
+	}
+	return cx.stack[0].Uint != 0, nil
```

Comment:
why do we expect non-zero "return code" on the stack? looks a bit counter-intuitive compare to process exit codes and go's err == nil approach

Reply:
I started thinking of this in terms of a logic expression, like `(A && B) || (C && !D)`, and when all that is evaluated it boils down to one true/false value for whether the transaction is approved.